File: src/main/java/com/indiepost/model/Post.java
package com.indiepost.model;

import com.indiepost.enums.Types;

import javax.persistence.*;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by jake on 7/24/16.
 */
@Entity
@Table(name = "Posts")
public class Post implements Serializable {

    private static final long serialVersionUID = -1121960490475976481L;

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "originalId")
    private Post original;

    @Column(name = "originalId", nullable = false, insertable = false, updatable = false)
    private Long originalId;

    @Column(nullable = false)
    private boolean featured = false;

    @Column(nullable = false)
    private boolean picked = false;

    @Column(nullable = false)
    private boolean splash = false;

    @Column(nullable = false)
    @Size(max = 100)
    private String title = "No Title";

    @Column(nullable = false, columnDefinition = "LONGTEXT")
    private String content = "";

    @Column(nullable = false)
    @Size(max = 400)
    private String excerpt = "";

    @Column(nullable = false)
    @Size(max = 30)
    private String displayName = "Indiepost";

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime modifiedAt;

    @Column(nullable = false)
    private LocalDateTime publishedAt;

    @Column(nullable = false)
    private boolean showLastUpdated = false;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "titleImageId")
    private ImageSet titleImage;

    @Column(name = "titleImageId", insertable = false, updatable = false)
    private Long titleImageId;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private Types.PostStatus status;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "authorId", nullable = false)
    private User author;

    @Column(name = "authorId", nullable = false, insertable = false, updatable = false)
    private Long authorId;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "editorId", nullable = false)
    private User editor;

    @Column(name = "editorId", nullable = false, insertable = false, updatable = false)
    private Long editorId;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "categoryId", nullable = false)
    private Category category;

    @Column(name = "categoryId", nullable = false, insertable = false, updatable = false)
    private Long categoryId;

    @OneToMany(
            mappedBy = "post",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @OrderBy("priority")
    private List<PostTag> postTags = new ArrayList<>();

    @OneToMany(
            mappedBy = "post",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @OrderBy("priority")
    private List<PostContributor> postContributors = new ArrayList<>();

    @OneToMany(
            mappedBy = "post",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    private List<PostReading> postReadings = new ArrayList<>();

    @OneToMany(
            mappedBy = "post",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    private List<Bookmark> postBookmarks = new ArrayList<>();

    @OneToMany(
            mappedBy = "post",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @OrderBy("priority")
    private List<PostPost> postPosts = new ArrayList<>();

    public List<PostReading> getPostReadings() {
        return postReadings;
    }

    public void setPostReadings(List<PostReading> postReadings) {
        this.postReadings = postReadings;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public ImageSet getTitleImage() {
        return titleImage;
    }

    public void setTitleImage(ImageSet titleImage) {
        this.titleImage = titleImage;
    }

    public Long getTitleImageId() {
        return titleImageId;
    }

    public void setTitleImageId(Long titleImageId) {
        this.titleImageId = titleImageId;
    }

    public String getDisplayName() {
        return displayName;
    }

    public void setDisplayName(String displayName) {
        this.displayName = displayName;
    }

    public LocalDateTime getModifiedAt() {
        return modifiedAt;
    }

    public void setModifiedAt(LocalDateTime modifiedAt) {
        this.modifiedAt = modifiedAt;
    }

    public LocalDateTime getPublishedAt() {
        return publishedAt;
    }

    public void setPublishedAt(LocalDateTime publishedAt) {
        this.publishedAt = publishedAt;
    }

    public String getExcerpt() {
        return excerpt;
    }

    public void setExcerpt(String excerpt) {
        this.excerpt = excerpt;
    }

    public Types.PostStatus getStatus() {
        return status;
    }

    public void setStatus(Types.PostStatus status) {
        this.status = status;
    }

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public void addTag(Tag tag, int priority) {
        PostTag postTag = new PostTag(this, tag, priority);
        this.postTags.add(postTag);
        tag.getPostTags().add(postTag);
    }

    public void removeTag(Tag tag) {
        for (Iterator<PostTag> iterator = postTags.iterator();
             iterator.hasNext(); ) {
            PostTag postTag = iterator.next();

            if (postTag.getPost().equals(this) &&
                    postTag.getTag().equals(tag)) {
                iterator.remove();
                postTag.getTag().getPostTags().remove(postTag);
                postTag.setPost(null);
                postTag.setTag(null);
            }
        }
    }

    public List<Tag> getTags() {
        return postTags.stream()
                .map(postTag -> postTag.getTag())
                .collect(Collectors.toList());
    }

    public void clearTags() {
        this.postTags.clear();
    }

    public void setPostContributors(List<PostContributor> postContributors) {
        this.postContributors = postContributors;
    }

    public void addContributor(Contributor contributor, int priority) {
        PostContributor postContributor =
                new PostContributor(this, contributor, priority);
        this.postContributors.add(postContributor);
        contributor.getPostContributors().add(postContributor);
    }

    public void removeContributor(Contributor contributor) {
        for (Iterator<PostContributor> iterator = postContributors.iterator();
             iterator.hasNext(); ) {
            PostContributor postContributor = iterator.next();

            if (postContributor.getPost().equals(this) &&
                    postContributor.getContributor().equals(contributor)) {
                iterator.remove();
                postContributor.getContributor().getPostContributors().remove(postContributor);
                postContributor.setPost(null);
                postContributor.setContributor(null);
            }
        }
    }

    public List<Contributor> getContributors() {
        return postContributors.stream()
                .map(postContributor -> postContributor.getContributor())
                .collect(Collectors.toList());
    }

    public void clearContributors() {
        this.postContributors.clear();
    }

    public void addRelatedPost(Post post, int priority) {
        PostPost postPost =
                new PostPost(this, post, priority);
        this.postPosts.add(postPost);
    }

    public void removeRelatedPost(Post post) {
        for (PostPost postPost : postPosts) {
            if (postPost.getPost().equals(this) &&
                    postPost.getRelatedPost().equals(post)) {
                postPosts.remove(postPost);
                postPost.getRelatedPost().getPostPosts().remove(this);
                postPost.setPost(null);
                postPost.setRelatedPost(null);
            }
        }
    }

    public List<Post> getPostPosts() {
        return postPosts.stream()
                .map(postPost -> postPost.getRelatedPost())
                .collect(Collectors.toList());
    }

    public void setPostPosts(List<PostPost> postPosts) {
        this.postPosts = postPosts;
    }

    public void clearRelatedPosts() {
        this.postPosts.clear();
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public User getAuthor() {
        return author;
    }

    public void setAuthor(User author) {
        this.author = author;
    }

    public User getEditor() {
        return editor;
    }

    public void setEditor(User editor) {
        this.editor = editor;
    }

    public Post getOriginal() {
        return original;
    }

    public void setOriginal(Post original) {
        this.original = original;
    }

    public Long getOriginalId() {
        return originalId;
    }

    public void setOriginalId(Long originalId) {
        this.originalId = originalId;
    }

    public Long getAuthorId() {
        return authorId;
    }

    public void setAuthorId(Long authorId) {
        this.authorId = authorId;
    }

    public Long getEditorId() {
        return editorId;
    }

    public void setEditorId(Long editorId) {
        this.editorId = editorId;
    }

    public Long getCategoryId() {
        return categoryId;
    }

    public void setCategoryId(Long categoryId) {
        this.categoryId = categoryId;
    }

    public boolean isFeatured() {
        return featured;
    }

    public void setFeatured(boolean featured) {
        this.featured = featured;
    }

    public boolean isPicked() {
        return picked;
    }

    public void setPicked(boolean picked) {
        this.picked = picked;
    }

    public boolean isSplash() {
        return splash;
    }

    public void setSplash(boolean splash) {
        this.splash = splash;
    }

    public List<Bookmark> getPostBookmarks() {
        return postBookmarks;
    }

    public void setPostBookmarks(List<Bookmark> postBookmarks) {
        this.postBookmarks = postBookmarks;
    }

    public boolean isShowLastUpdated() {
        return showLastUpdated;
    }

    public void setShowLastUpdated(boolean showLastUpdated) {
        this.showLastUpdated = showLastUpdated;
    }
}

File: src/main/kotlin/com/indiepost/model/Post.kt
package com.indiepost.model

import com.indiepost.enums.Types
import java.time.LocalDateTime
import java.util.*
import java.util.stream.Collectors
import javax.persistence.*
import javax.validation.constraints.Size

/**
 * Created by jake on 7/24/16.
 */
@Entity
@Table(name = "Posts")
data class Post(
        @Id
        @GeneratedValue(strategy = GenerationType.AUTO)
        var id: Long? = null,

        @Column(nullable = false)
        var isFeatured: Boolean = false,

        @Column(nullable = false)
        var isPicked: Boolean = false,

        @Column(nullable = false)
        var isSplash: Boolean = false,

        @Column(nullable = false)
        @Size(max = 100)
        var title: String = "No Title",

        @Column(nullable = false, columnDefinition = "LONGTEXT")
        var content: String = "",

        @Column(nullable = false)
        @Size(max = 400)
        var excerpt: String = "",

        @Column(nullable = false)
        @Size(max = 30)
        var displayName: String = "INDIEPOST",

        @Column(nullable = false)
        var createdAt: LocalDateTime? = null,

        @Column(nullable = false)
        var modifiedAt: LocalDateTime? = null,

        @Column(nullable = false)
        var publishedAt: LocalDateTime? = null,

        @Column(nullable = false)
        var isShowLastUpdated: Boolean = false,

        @Column(nullable = false)
        @Enumerated(EnumType.STRING)
        var status: Types.PostStatus? = null
) {
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "originalId")
    var original: Post? = null

    @Column(name = "originalId", nullable = false, insertable = false, updatable = false)
    var originalId: Long? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "titleImageId")
    var titleImage: ImageSet? = null

    @Column(name = "titleImageId", insertable = false, updatable = false)
    var titleImageId: Long? = null

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "authorId", nullable = false)
    var author: User? = null

    @Column(name = "authorId", nullable = false, insertable = false, updatable = false)
    var authorId: Long? = null

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "editorId", nullable = false)
    var editor: User? = null

    @Column(name = "editorId", nullable = false, insertable = false, updatable = false)
    var editorId: Long? = null

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "categoryId", nullable = false)
    var category: Category? = null

    @Column(name = "categoryId", nullable = false, insertable = false, updatable = false)
    var categoryId: Long? = null

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "tagId", nullable = false)
    var primaryTag: Tag? = null

    @Column(name = "tagId", nullable = false, insertable = false, updatable = false)
    var primaryTagId: Long? = null

    @OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    @OrderBy("priority")
    private val postTags: MutableList<PostTag> = ArrayList()

    @OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    @OrderBy("priority")
    var postProfile: MutableList<PostProfile> = ArrayList()

    @OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    var postReadings: MutableList<PostReading> = ArrayList()

    @OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    var postBookmarks: MutableList<Bookmark> = ArrayList()

    @OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    @OrderBy("priority")
    private var postPosts: MutableList<PostPost> = ArrayList()

    val tags: MutableList<Tag>
        get() = postTags.stream()
                .map { postTag -> postTag.tag }
                .collect(Collectors.toList())

    val profiles: MutableList<Profile>
        get() = postProfile.stream()
                .map { postProfile -> postProfile.profile }
                .collect(Collectors.toList())

    fun addTag(tag: Tag, priority: Int) {
        val postTag = PostTag(this, tag, priority)
        this.postTags.add(postTag)
        tag.postTags.add(postTag)
    }

    fun removeTag(tag: Tag) {
        val iterator = postTags.iterator()
        while (iterator.hasNext()) {
            val postTag = iterator.next()

            if (postTag.post == this && postTag.tag == tag) {
                iterator.remove()
                postTag.tag?.let {
                    it.postTags.remove(postTag)
                }
                postTag.post = null
                postTag.tag = null
            }
        }
    }

    fun clearTags() {
        this.postTags.clear()
    }


    fun addProfile(profile: Profile, priority: Int) {
        val postProfile = PostProfile(this, profile, priority)
        this.postProfile.add(postProfile)
        profile.postProfile.add(postProfile)
    }

    fun removeProfile(profile: Profile) {
        val iterator = postProfile.iterator()
        while (iterator.hasNext()) {
            val postProfile = iterator.next()

            if (postProfile.post == this && postProfile.profile == profile) {
                iterator.remove()
                postProfile.profile?.let {
                    it.postProfile.remove(postProfile)
                }
                postProfile.post = null
                postProfile.profile = null
            }
        }
    }

    fun clearProfiles() {
        this.postProfile.clear()
    }

    fun addRelatedPost(post: Post, priority: Int) {
        val postPost = PostPost(this, post, priority)
        this.postPosts.add(postPost)
    }

    fun removeRelatedPost(post: Post) {
        for (postPost in postPosts) {
            if (postPost.post == this && postPost.relatedPost == post) {
                postPosts.remove(postPost)
                postPost.post = null
                postPost.relatedPost = null
            }
        }
    }

    fun getPostPosts(): MutableList<Post> {
        return postPosts.stream()
                .map { postPost -> postPost.relatedPost }
                .collect(Collectors.toList())
    }

    fun setPostPosts(postPosts: MutableList<PostPost>) {
        this.postPosts = postPosts
    }

    fun clearRelatedPosts() {
        this.postPosts.clear()
    }

    companion object {
        private const val serialVersionUID = -1121960490475976481L
    }
}